merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,1211 @@
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use ark_bn254::Fr;
+    use ark_ff::UniformRand;
+    use rand::{thread_rng, Rng};
+
+    use super::*;
+
+    #[test]
+    fn test_find_next_prime() {
+        assert_eq!(find_next_prime(0.0), 2.0);
+        assert_eq!(find_next_prime(2.0), 2.0);
+        assert_eq!(find_next_prime(3.0), 3.0);
+        assert_eq!(find_next_prime(4.0), 5.0);
+
+        assert_eq!(find_next_prime(10.0), 11.0);
+        assert_eq!(find_next_prime(28.0), 29.0);
+
+        assert_eq!(find_next_prime(100.0), 101.0);
+        assert_eq!(find_next_prime(1000.0), 1009.0);
+
+        assert_eq!(find_next_prime(102.0), 103.0);
+        assert_eq!(find_next_prime(105.0), 107.0);
+
+        assert_eq!(find_next_prime(7900.0), 7901.0);
+        assert_eq!(find_next_prime(7907.0), 7907.0);
+    }
+
+    #[test]
+    fn test_capacity_cells() {
+        assert_eq!(HashSet::<u16>::capacity_indices(256, 0.5).unwrap(), 521.0);
+        assert_eq!(HashSet::<u16>::capacity_indices(4800, 0.7).unwrap(), 6857.0);
+    }
+
+    /// Manual test cases. A simple check whether basic properties of the hash
+    /// set work.
+    #[test]
+    fn test_hash_set_manual() {
+        let mut hs = HashSet::<u16>::new(521, 256, 4).unwrap();
+
+        // Insert an element and immediately mark it with a sequence number.
+        // An equivalent to a single insertion in Light Protocol
+        let element_1_1 = 1.to_biguint().unwrap();
+        hs.insert(&element_1_1, 0).unwrap();
+        hs.mark_with_sequence_number(&element_1_1, 1).unwrap();
+
+        // Check if element exists in the set.
+        assert_eq!(hs.contains(&element_1_1, 1).unwrap(), true);
+        // Try inserting the same element, even though we didn't reach the
+        // threshold.
+        assert!(matches!(
+            hs.insert(&element_1_1, 1),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+
+        // Insert multiple elements and mark them with one sequence number.
+        // An equivalent to a batched insertion in Light Protocol.
+
+        let element_2_3 = 3.to_biguint().unwrap();
+        let element_2_6 = 6.to_biguint().unwrap();
+        let element_2_8 = 8.to_biguint().unwrap();
+        let element_2_9 = 9.to_biguint().unwrap();
+        hs.insert(&element_2_3, 1).unwrap();
+        hs.insert(&element_2_6, 1).unwrap();
+        hs.insert(&element_2_8, 1).unwrap();
+        hs.insert(&element_2_9, 1).unwrap();
+        assert_eq!(hs.contains(&element_2_3, 2).unwrap(), true);
+        assert_eq!(hs.contains(&element_2_6, 2).unwrap(), true);
+        assert_eq!(hs.contains(&element_2_8, 2).unwrap(), true);
+        assert_eq!(hs.contains(&element_2_9, 2).unwrap(), true);
+        hs.mark_with_sequence_number(&element_2_3, 2).unwrap();
+        hs.mark_with_sequence_number(&element_2_6, 2).unwrap();
+        hs.mark_with_sequence_number(&element_2_8, 2).unwrap();
+        hs.mark_with_sequence_number(&element_2_9, 2).unwrap();
```

Comment:
These method calls overwrite 2 with 2, in other words don't do anything.
```suggestion
        hs.mark_with_sequence_number(&element_2_3, 3).unwrap();
        hs.mark_with_sequence_number(&element_2_6, 3).unwrap();
        hs.mark_with_sequence_number(&element_2_8, 3).unwrap();
        hs.mark_with_sequence_number(&element_2_9, 3).unwrap();
```

Reply:
No, these elements weren't marked with sequence numbers before.

The sequence number provided in `insert` is the sequence number used to potentially vacate a cell. But it's not being assigned yet. New cells used by `insert` have `seq: None`. That's because we want to assign sequence numbers when inserting to the Merkle tree, not to the indexed array